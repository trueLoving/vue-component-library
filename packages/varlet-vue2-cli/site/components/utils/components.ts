import Vue, { ComponentOptions } from 'vue'
import { CombinedVueInstance } from 'vue/types/vue'

export function pickProps<T, U extends keyof T>(props: T, propsKey: U): T[U]
export function pickProps<T, U extends keyof T>(props: T, propsKey: U[]): Pick<T, U>
export function pickProps(props: any, propsKey: any): any {
  return Array.isArray(propsKey)
    ? propsKey.reduce((pickedProps: any, key) => {
        pickedProps[key] = props[key]
        return pickedProps
      }, {})
    : props[propsKey]
}

export interface MountComponentApi {
  instance: CombinedVueInstance<any, any, any, any, any>
  unmount(): void
}

export function mountComponent(component: ComponentOptions<Vue>, container = 'body', options = {}): MountComponentApi {
  const instance = new (Vue.extend(component))(options)
  const el = instance.$mount().$el
  const wrapper = document.querySelector(container)!
  wrapper.appendChild(el)

  return {
    instance,
    unmount() {
      instance.$destroy()
      wrapper.removeChild(el)
    },
  }
}
