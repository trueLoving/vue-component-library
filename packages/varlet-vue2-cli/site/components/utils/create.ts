import Vue, { ComponentOptions, RenderContext, VueConstructor } from 'vue'
import { SlotsMixin } from './mixins/slots'
import { ListenersMixin } from './mixins/listeners'

export type ComponentOptionsPlugin = ComponentOptions<Vue> & { install?: (app: VueConstructor) => void }

export function nextTick() {
  return new Promise(Vue.nextTick)
}

export function mergeSlots(context: RenderContext) {
  const scopedSlots = context.scopedSlots ?? context.data.scopedSlots ?? {}
  const slots = context.slots()

  Object.keys(slots).forEach((key) => {
    if (!scopedSlots[key]) {
      scopedSlots[key] = () => slots[key]
    }
  })

  return scopedSlots
}

export function defineComponent(sfc: ComponentOptionsPlugin & { name: string }) {
  if (sfc.mixins) {
    sfc.mixins.push(SlotsMixin)
    sfc.mixins.push(ListenersMixin)
  } else {
    sfc.mixins = [SlotsMixin, ListenersMixin]
  }

  return sfc
}
