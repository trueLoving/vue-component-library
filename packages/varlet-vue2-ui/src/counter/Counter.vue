<template>
  <div class="var-counter var--box">
    <div
      class="var-counter__controller var-elevation--2"
      :class="[disabled || formDisabled ? 'var-counter--disabled' : null, errorMessage ? 'var-counter--error' : null]"
      :style="{ background: color ? color : undefined }"
      v-bind="$attrs"
    >
      <var-icon
        class="var-counter__decrement-button"
        var-counter-cover
        name="minus"
        :class="[!decrementButton ? 'var-counter--hidden' : null]"
        :style="{
          width: toSizeUnit(buttonSize),
          height: toSizeUnit(buttonSize),
        }"
        v-ripple="{
          disabled: !ripple || disabled || readonly || disableDecrement || !decrementButton || isMin,
        }"
        @click="decrement"
        @touchstart.native="pressDecrement"
        @touchend.native="releaseDecrement"
        @touchcancel.native="releaseDecrement"
      />
      <input
        class="var-counter__input"
        :style="{
          width: toSizeUnit(inputWidth),
          fontSize: toSizeUnit(inputTextSize),
        }"
        :inputmode="toNumber(decimalLength) === 0 ? 'numeric' : 'decimal'"
        :readonly="readonly || formReadonly"
        :disabled="disabled || formDisabled || disableInput"
        v-model="inputValue"
        @change="handleChange"
      />
      <var-icon
        class="var-counter__increment-button"
        var-counter-cover
        name="plus"
        :class="[!incrementButton ? 'var-counter--hidden' : null]"
        :style="{
          width: toSizeUnit(buttonSize),
          height: toSizeUnit(buttonSize),
        }"
        v-ripple="{
          disabled: !ripple || disabled || readonly || disableIncrement || !incrementButton || isMax,
        }"
        @click="increment"
        @touchstart.native="pressIncrement"
        @touchend.native="releaseIncrement"
        @touchcancel.native="releaseIncrement"
      />
    </div>

    <var-form-details :error-message="errorMessage"></var-form-details>
  </div>
</template>

<script>
import VarIcon from '../icon'
import VarFormDetails from '../form-details'
import Ripple from '../ripple'
import { defineComponent } from '../utils/create'
import { Decimal } from 'decimal.js'
import { props } from './props'
import { toNumber } from '../utils/shared'
import { toSizeUnit } from '../utils/elements'
import { ValidationMixin } from '../utils/mixins/validation'
import { createChildrenMixin } from '../utils/mixins/relation'

const SPEED = 100
const DELAY = 600

export default defineComponent({
  name: 'VarCounter',

  components: {
    VarIcon,
    VarFormDetails,
  },

  directives: { Ripple },

  mixins: [ValidationMixin, createChildrenMixin('form', { parentKey: 'form', childrenKey: 'formItems' })],

  inheritAttrs: false,

  props,

  data: () => ({
    inputValue: '',
    incrementTimer: null,
    decrementTimer: null,
    incrementDelayTimer: null,
    decrementDelayTimer: null,
  }),

  computed: {
    isMax() {
      const { max, value } = this
      return max != null && toNumber(value) >= toNumber(max)
    },

    isMin() {
      const { min, value } = this
      return min != null && toNumber(value) <= toNumber(min)
    },

    formDisabled() {
      return this.form?.disabled
    },

    formReadonly() {
      return this.form?.readonly
    },
  },

  created() {
    this.setNormalizedValue(this.normalizeValue(String(this.value)))
  },

  watch: {
    value(newValue) {
      this.setNormalizedValue(this.normalizeValue(String(newValue)))
      this.getListeners().onChange?.(toNumber(newValue))
    },
  },

  methods: {
    // expose
    reset() {
      const { min, getListeners } = this

      getListeners().onInput?.(min != null ? toNumber(min) : 0)
      this.resetValidation()
    },

    toNumber,

    toSizeUnit,

    validate() {
      this._validate(this.rules, this.value)
    },

    validateWithTrigger(trigger) {
      this.$nextTick(() => {
        const { validateTrigger, rules, value } = this
        this._validateWithTrigger(validateTrigger, trigger, rules, value)
      })
    },

    normalizeValue(value) {
      const { decimalLength, max, min } = this
      let num = toNumber(value)

      if (max != null && num > toNumber(max)) {
        num = toNumber(max)
      }

      if (min != null && num < toNumber(min)) {
        num = toNumber(min)
      }

      value = String(num)

      if (decimalLength != null) {
        value = num.toFixed(toNumber(decimalLength))
      }

      return value
    },

    handleChange(event) {
      const { lazyChange, getListeners, change, validateWithTrigger } = this
      const { value } = event.target
      const normalizedValue = this.normalizeValue(value)

      lazyChange
        ? getListeners().onBeforeChange?.(toNumber(normalizedValue), change)
        : this.setNormalizedValue(normalizedValue)

      validateWithTrigger('onInputChange')
    },

    decrement() {
      const { disabled, readonly, disableDecrement, decrementButton, lazyChange, step, value, getListeners } = this

      const { onDecrement, onBeforeChange } = getListeners()

      if (this.formDisabled || this.formReadonly || disabled || readonly || disableDecrement || !decrementButton) {
        return
      }

      if (this.isMin) {
        return
      }

      const localValue = new Decimal(toNumber(value)).minus(new Decimal(toNumber(step))).toString()
      const normalizedValue = this.normalizeValue(localValue)
      const normalizedValueNum = toNumber(normalizedValue)
      onDecrement?.(normalizedValueNum)

      if (lazyChange) {
        onBeforeChange?.(normalizedValueNum, this.change)
      } else {
        this.setNormalizedValue(normalizedValue)
        this.validateWithTrigger('onDecrement')
      }
    },

    increment() {
      const { disabled, readonly, disableIncrement, incrementButton, lazyChange, step, value, getListeners } = this

      const { onIncrement, onBeforeChange } = getListeners()

      if (this.formDisabled || this.formReadonly || disabled || readonly || disableIncrement || !incrementButton) {
        return
      }

      if (this.isMax) {
        return
      }

      const localValue = new Decimal(toNumber(value)).plus(new Decimal(toNumber(step))).toString()
      const normalizedValue = this.normalizeValue(localValue)
      const normalizedValueNum = toNumber(normalizedValue)

      onIncrement?.(normalizedValueNum)

      if (lazyChange) {
        onBeforeChange?.(normalizedValueNum, this.change)
      } else {
        this.setNormalizedValue(normalizedValue)
        this.validateWithTrigger('onIncrement')
      }
    },

    pressDecrement() {
      const { press, lazyChange } = this

      if (!press || lazyChange) {
        return
      }

      this.decrementDelayTimer = window.setTimeout(() => {
        this.continuedDecrement()
      }, DELAY)
    },

    pressIncrement() {
      const { press, lazyChange } = this

      if (!press || lazyChange) {
        return
      }

      this.incrementDelayTimer = window.setTimeout(() => {
        this.continuedIncrement()
      }, DELAY)
    },

    releaseDecrement() {
      this.decrementTimer && clearTimeout(this.decrementTimer)
      this.decrementDelayTimer && clearTimeout(this.decrementDelayTimer)
    },

    releaseIncrement() {
      this.incrementTimer && clearTimeout(this.incrementTimer)
      this.incrementDelayTimer && clearTimeout(this.incrementDelayTimer)
    },

    continuedIncrement() {
      this.incrementTimer = window.setTimeout(() => {
        this.increment()
        this.continuedIncrement()
      }, SPEED)
    },

    continuedDecrement() {
      this.decrementTimer = window.setTimeout(() => {
        this.decrement()
        this.continuedDecrement()
      }, SPEED)
    },

    setNormalizedValue(normalizedValue) {
      this.inputValue = normalizedValue

      const normalizedValueNum = toNumber(normalizedValue)

      this.getListeners().onInput?.(normalizedValueNum)
    },

    change(value) {
      this.setNormalizedValue(this.normalizeValue(String(value)))
      this.validateWithTrigger('onLazyChange')
    },
  },
})
</script>

<style lang="less">
@import '../styles/common';
@import '../styles/elevation';
@import '../icon/icon';
@import '../form-details/formDetails';
@import '../ripple/ripple';
@import './counter';
</style>
