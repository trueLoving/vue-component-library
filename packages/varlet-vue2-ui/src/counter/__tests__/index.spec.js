import example from '../example'
import Counter from '..'
import VarCounter from '../Counter'
import { mount } from '@vue/test-utils'
import Vue from 'vue'
import { delay, trigger } from '../../utils/jest'

test('test counter example', () => {
  const wrapper = mount(example)
  expect(wrapper.html()).toMatchSnapshot()
  wrapper.destroy()
})

test('test counter plugin', () => {
  Vue.use(Counter)
  expect(Vue.component(Counter.name)).toBeTruthy()
})

test('test counter increment & decrement', async () => {
  const wrapper = mount({
    components: {
      [VarCounter.name]: VarCounter,
    },
    data: () => ({
      value: 0,
    }),
    template: `<var-counter v-model="value" />`,
  })

  await wrapper.find('.var-counter__increment-button').trigger('click')
  expect(wrapper.vm.value).toBe(1)
  await wrapper.find('.var-counter__decrement-button').trigger('click')
  expect(wrapper.vm.value).toBe(0)

  wrapper.destroy()
})

test('test counter initial value over max', async () => {
  const wrapper = mount({
    components: {
      [VarCounter.name]: VarCounter,
    },
    data: () => ({
      value: 11,
    }),
    template: `<var-counter :max="10" v-model="value" />`,
  })
  expect(wrapper.vm.value).toBe(10)

  wrapper.destroy()
})

test('test counter initial value less min', async () => {
  const Wrapper = {
    components: {
      [VarCounter.name]: VarCounter,
    },
    data: () => ({
      value: -1,
    }),
    template: `<var-counter :min="0" v-model="value" />`,
  }
  const wrapper = mount(Wrapper)
  expect(wrapper.vm.value).toBe(0)

  wrapper.destroy()
})

test('test counter onChange', async () => {
  const onChange = jest.fn()
  const Wrapper = {
    components: {
      [VarCounter.name]: VarCounter,
    },
    data: () => ({
      value: 0,
    }),
    methods: {
      onChange,
    },
    template: `<var-counter v-model="value" @change="onChange" />`,
  }
  const wrapper = mount(Wrapper)

  wrapper.find('.var-counter__input').setValue('1')
  await wrapper.find('.var-counter__input').trigger('change')
  expect(onChange).lastCalledWith(1)

  wrapper.destroy()
})

test('test counter press increment', async () => {
  const wrapper = mount({
    components: {
      [VarCounter.name]: VarCounter,
    },
    data: () => ({
      value: 0,
    }),
    template: `<var-counter v-model="value" />`,
  })

  await trigger(wrapper.find('.var-counter__increment-button'), 'touchstart')
  await delay(800)

  const current = wrapper.vm.value
  expect(current).toBeGreaterThan(0)

  await trigger(wrapper.find('.var-counter__increment-button'), 'touchend')
  await delay(100)
  expect(wrapper.vm.value).toBe(current)

  wrapper.destroy()
})

test('test counter press decrement', async () => {
  const wrapper = mount({
    components: {
      [VarCounter.name]: VarCounter,
    },
    data: () => ({
      value: 0,
    }),
    template: `<var-counter v-model="value" />`,
  })

  await trigger(wrapper.find('.var-counter__decrement-button'), 'touchstart')
  await delay(750)

  const current = wrapper.vm.value
  expect(current).toBeLessThan(0)

  await trigger(wrapper.find('.var-counter__decrement-button'), 'touchend')
  await delay(750)
  expect(wrapper.vm.value).toBe(current)

  wrapper.destroy()
})

test('test counter lazy change', async () => {
  const wrapper = mount({
    components: {
      [VarCounter.name]: VarCounter,
    },
    data: () => ({
      value: 0,
    }),
    methods: {
      onBeforeChange(value, change) {
        const isInc = value > wrapper.vm.value
        change(isInc ? value + 1 : value - 1)
      },
    },
    template: `<var-counter lazy-change v-model="value" @before-change="onBeforeChange" />`,
  })

  await wrapper.find('.var-counter__increment-button').trigger('click')
  expect(wrapper.vm.value).toBe(2)

  await wrapper.find('.var-counter__decrement-button').trigger('click')
  expect(wrapper.vm.value).toBe(0)

  wrapper.destroy()
})

test('test counter disabled', async () => {
  const wrapper = mount({
    components: {
      [VarCounter.name]: VarCounter,
    },
    data: () => ({
      value: 0,
    }),
    template: `<var-counter disabled v-model="value" />`,
  })

  await wrapper.find('.var-counter__increment-button').trigger('click')
  expect(wrapper.vm.value).toBe(0)

  await wrapper.find('.var-counter__decrement-button').trigger('click')
  expect(wrapper.vm.value).toBe(0)
})

test('test counter validation', async () => {
  const wrapper = mount({
    components: {
      [VarCounter.name]: VarCounter,
    },
    data: () => ({
      value: 0,
    }),
    template: `<var-counter ref="counter" :rules="[v => v > 0 || '必须大于0']" v-model="value" />`,
  })

  const { counter } = wrapper.vm.$refs

  counter.validate()
  await delay(16)

  expect(wrapper.html()).toMatchSnapshot()
  expect(wrapper.find('.var-form-details__message').text()).toBe('必须大于0')

  await wrapper.find('.var-counter__increment-button').trigger('click')
  await delay(16)
  expect(wrapper.find('.var-form-details__message').exists()).toBeFalsy()

  counter.reset()
  await delay(16)
  expect(wrapper.vm.value).toBe(0)
})
