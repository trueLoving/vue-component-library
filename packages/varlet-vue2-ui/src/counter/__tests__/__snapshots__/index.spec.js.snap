// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test counter example 1`] = `
"<div class=\\"var-counter-example\\">
  <div class=\\"app-type\\">基本使用</div>
  <div class=\\"var-counter var--box\\">
    <div class=\\"var-counter__controller var-elevation--2\\"><i class=\\"var-icon var-counter__decrement-button var-icon--set var-icon-minus\\" style=\\"transition: transform 0ms;\\" var-counter-cover=\\"\\"></i> <input inputmode=\\"numeric\\" class=\\"var-counter__input\\"> <i class=\\"var-icon var-counter__increment-button var-icon--set var-icon-plus\\" style=\\"transition: transform 0ms;\\" var-counter-cover=\\"\\"></i></div>
    <transition-stub name=\\"var-form-details\\">
      <!---->
    </transition-stub>
  </div>
  <div class=\\"app-type\\">设置取值范围</div>
  <div class=\\"var-counter var--box\\">
    <div class=\\"var-counter__controller var-elevation--2\\"><i class=\\"var-icon var-counter__decrement-button var-icon--set var-icon-minus\\" style=\\"transition: transform 0ms;\\" var-counter-cover=\\"\\"></i> <input inputmode=\\"numeric\\" class=\\"var-counter__input\\"> <i class=\\"var-icon var-counter__increment-button var-icon--set var-icon-plus\\" style=\\"transition: transform 0ms;\\" var-counter-cover=\\"\\"></i></div>
    <transition-stub name=\\"var-form-details\\">
      <!---->
    </transition-stub>
  </div>
  <div class=\\"app-type\\">设置步长</div>
  <div class=\\"var-counter var--box\\">
    <div class=\\"var-counter__controller var-elevation--2\\"><i class=\\"var-icon var-counter__decrement-button var-icon--set var-icon-minus\\" style=\\"transition: transform 0ms;\\" var-counter-cover=\\"\\"></i> <input inputmode=\\"numeric\\" class=\\"var-counter__input\\"> <i class=\\"var-icon var-counter__increment-button var-icon--set var-icon-plus\\" style=\\"transition: transform 0ms;\\" var-counter-cover=\\"\\"></i></div>
    <transition-stub name=\\"var-form-details\\">
      <!---->
    </transition-stub>
  </div>
  <div class=\\"app-type\\">保留小数</div>
  <div class=\\"var-counter var--box\\">
    <div class=\\"var-counter__controller var-elevation--2\\"><i class=\\"var-icon var-counter__decrement-button var-icon--set var-icon-minus\\" style=\\"transition: transform 0ms;\\" var-counter-cover=\\"\\"></i> <input inputmode=\\"decimal\\" class=\\"var-counter__input\\"> <i class=\\"var-icon var-counter__increment-button var-icon--set var-icon-plus\\" style=\\"transition: transform 0ms;\\" var-counter-cover=\\"\\"></i></div>
    <transition-stub name=\\"var-form-details\\">
      <!---->
    </transition-stub>
  </div>
  <div class=\\"app-type\\">设置尺寸</div>
  <div class=\\"var-counter var--box\\">
    <div class=\\"var-counter__controller var-elevation--2\\"><i class=\\"var-icon var-counter__decrement-button var-icon--set var-icon-minus\\" style=\\"transition: transform 0ms; width: 36px; height: 36px;\\" var-counter-cover=\\"\\"></i> <input inputmode=\\"numeric\\" class=\\"var-counter__input\\" style=\\"width: 50px; font-size: 18px;\\"> <i class=\\"var-icon var-counter__increment-button var-icon--set var-icon-plus\\" style=\\"transition: transform 0ms; width: 36px; height: 36px;\\" var-counter-cover=\\"\\"></i></div>
    <transition-stub name=\\"var-form-details\\">
      <!---->
    </transition-stub>
  </div>
  <div class=\\"app-type\\">禁用</div>
  <div class=\\"var-counter var--box\\">
    <div class=\\"var-counter__controller var-elevation--2 var-counter--disabled\\"><i class=\\"var-icon var-counter__decrement-button var-icon--set var-icon-minus\\" style=\\"transition: transform 0ms;\\" var-counter-cover=\\"\\"></i> <input inputmode=\\"numeric\\" disabled=\\"disabled\\" class=\\"var-counter__input\\"> <i class=\\"var-icon var-counter__increment-button var-icon--set var-icon-plus\\" style=\\"transition: transform 0ms;\\" var-counter-cover=\\"\\"></i></div>
    <transition-stub name=\\"var-form-details\\">
      <!---->
    </transition-stub>
  </div>
  <div class=\\"app-type\\">只读</div>
  <div class=\\"var-counter var--box\\">
    <div class=\\"var-counter__controller var-elevation--2\\"><i class=\\"var-icon var-counter__decrement-button var-icon--set var-icon-minus\\" style=\\"transition: transform 0ms;\\" var-counter-cover=\\"\\"></i> <input inputmode=\\"numeric\\" readonly=\\"readonly\\" class=\\"var-counter__input\\"> <i class=\\"var-icon var-counter__increment-button var-icon--set var-icon-plus\\" style=\\"transition: transform 0ms;\\" var-counter-cover=\\"\\"></i></div>
    <transition-stub name=\\"var-form-details\\">
      <!---->
    </transition-stub>
  </div>
  <div class=\\"app-type\\">异步变更</div>
  <div class=\\"var-counter var--box\\">
    <div class=\\"var-counter__controller var-elevation--2\\"><i class=\\"var-icon var-counter__decrement-button var-icon--set var-icon-minus\\" style=\\"transition: transform 0ms;\\" var-counter-cover=\\"\\"></i> <input inputmode=\\"numeric\\" class=\\"var-counter__input\\"> <i class=\\"var-icon var-counter__increment-button var-icon--set var-icon-plus\\" style=\\"transition: transform 0ms;\\" var-counter-cover=\\"\\"></i></div>
    <transition-stub name=\\"var-form-details\\">
      <!---->
    </transition-stub>
  </div>
  <div class=\\"app-type\\">字段校验</div>
  <div class=\\"var-counter var--box\\">
    <div class=\\"var-counter__controller var-elevation--2\\"><i class=\\"var-icon var-counter__decrement-button var-icon--set var-icon-minus\\" style=\\"transition: transform 0ms;\\" var-counter-cover=\\"\\"></i> <input inputmode=\\"numeric\\" class=\\"var-counter__input\\"> <i class=\\"var-icon var-counter__increment-button var-icon--set var-icon-plus\\" style=\\"transition: transform 0ms;\\" var-counter-cover=\\"\\"></i></div>
    <transition-stub name=\\"var-form-details\\">
      <!---->
    </transition-stub>
  </div>
  <div class=\\"space\\"></div>
</div>"
`;

exports[`test counter validation 1`] = `
"<div class=\\"var-counter var--box\\">
  <div class=\\"var-counter__controller var-elevation--2 var-counter--error\\"><i class=\\"var-icon var-counter__decrement-button var-icon--set var-icon-minus\\" style=\\"transition: transform 0ms;\\" var-counter-cover=\\"\\"></i> <input inputmode=\\"numeric\\" class=\\"var-counter__input\\"> <i class=\\"var-icon var-counter__increment-button var-icon--set var-icon-plus\\" style=\\"transition: transform 0ms;\\" var-counter-cover=\\"\\"></i></div>
  <transition-stub name=\\"var-form-details\\">
    <div class=\\"var-form-details\\">
      <div class=\\"var-form-details__message\\">必须大于0</div>
      <div class=\\"var-form-details__length\\"></div>
    </div>
  </transition-stub>
</div>"
`;
